qiskit/algorithms/quantum_time_evolution/variational/variational_principles/variational_principle.py

```diff
@@ -0,0 +1,107 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Class for a Variational Principle."""
+
+from abc import ABC, abstractmethod
+from typing import Union, List, Callable
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import Parameter
+from qiskit.opflow import (
+    CircuitQFI,
+    CircuitGradient,
+    StateFn,
+    ListOp,
+    OperatorBase,
+)
+
+
+class VariationalPrinciple(ABC):
+    """Class for a Variational Principle."""
+
+    def __init__(
```

Comment:
I think there was some work done by @Zoufalc to ensure that these methods return both real and imaginary parts. Then we just select a relevant part in specializations. But I am open to discuss more, also in the context of SPSA techniques because perhaps not all the possibilities were considered.

Reply:
Actually, I misunderstood what this returns -- it does not return the general QFI tensor / gradient but just whatever matrix / vector the variational principle requires. 🙂